### PR TITLE
ci(github): migrate ci from circleci to github

### PR DESCRIPTION
**Pull request by rfprod**
- [x] reimplement circleci workflows using github actions ci;
- [x] clean up github workflows, add comments to justify suggestions;
- [x] implement a reusable github workflow that might be used later, add explanatory comments;
- [x] use a reusable workflow to setup all jobs;
- [x] use a matrix strategy to parallelize scripts execution;
- [x] `echo` explanation of the `find ... shred ...` command;
- [x] cache `@ngxs` build artifacts, add a reusable workflow to check for build artifacts;
- [x] cache `app/integrations` dists (needed for `bundlesize` job);
- [x] convert reusable workflows to composite actions;

## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/ngxs/store/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

Implementation of CI workflows using GitHub Actions CI.

```
[ ] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[x] CI related changes
[ ] Documentation content changes
[ ] Other... Please describe:
```

## What is the current behavior?

CircleCI does not always work as expected.

Issue Number: N/A

## What is the new behavior?

CircleCI workflows were reimplementd using GitHub Actions CI.

## Does this PR introduce a breaking change?

```
[ ] Yes
[x] No
```

## Other information

There're some TODOs that should be reviewed and addressed in the future.
GitHub Actions CI workflow steps that publish packages to NPM are commented out.
Before uncommenting the GitHub Actions CI workflow steps that publish packages to NPM, the new workflows must be tested first.

### CODE REVIEW

**Review by markwhitfeld — COMMENTED** [review]
Great start so far!
I think we should also leave the Circle CI stuff in place too so that we have 2 sets of CI running (because, why not 😉 ).
Some comments:
- I see quite a bit of repetition between these files, do github actions have a way of sharing steps, etc between multiple workflows (forgive my ignorance, and inexperience with GH actions if it shows).
- Is it possible to do the same as what we have in CircleCI, where the build happens, and then the tasks all happen mainly in parallel where they don't depend on each other?

**Comment by rfprod** [comment]
> Great start so far! I think we should also leave the Circle CI stuff in place too so that we have 2 sets of CI running (because, why not wink ). Some comments:
> 
> * I see quite a bit of repetition between these files, do github actions have a way of sharing steps, etc between multiple workflows (forgive my ignorance, and inexperience with GH actions if it shows).
> * Is it possible to do the same as what we have in CircleCI, where the build happens, and then the tasks all happen mainly in parallel where they don't depend on each other?

@markwhitfeld there are reusable workflows, but it might be a bit more complex to read and maintain it for someone inexperienced with GH actions. I will try to make it as simple as possible, need to check if its possible to share executions step results between reusable workflows. This feature `reusable workflows` was introduced relatively recently.
Some links:
- https://docs.github.com/en/actions/using-workflows/reusing-workflows
- https://github.blog/2022-02-10-using-reusable-workflows-github-actions/

**Comment by rfprod** [comment]
@markwhitfeld check my latest commit in this branch please.

File: .github/workflows/pr-validation.yml
Comment (markwhitfeld) [inline .github/workflows/pr-validation.yml:21]: Wouldn't it make more sense to reference the reusable workflow through an internal path for this head? Unless there is possibly a security risk in doing so?
I was thinking it would have been something like this:
```suggestion
  #   uses: ./.github/workflows/get-variables.yml
```

File: .github/workflows/pr-validation.yml
Comment (markwhitfeld) [inline .github/workflows/pr-validation.yml:82]: I would think that everything from this line and upwards would be a good candidate for a reusable workflow `setup-project.yml`.

File: .github/workflows/pr-validation.yml
Comment (markwhitfeld) [inline .github/workflows/pr-validation.yml:89]: I see that jobs run in parallel by default and you can control sequencing through the [needs](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idneeds) property.
I think steps can only be sequential.

It would be ideal if all the linting and testing steps after the build were other jobs (declaring a `needs: build` ) so that they can run in parallel, but wait for the build job first

I think that all of these testing/checking steps may be a good candidate for the use of the matrix job, where the script run is the varied part of the job.

File: .github/workflows/pr-validation.yml
Comment (markwhitfeld) [inline .github/workflows/pr-validation.yml:149]: I think that there is value to push the coverage to code climate so that the code climate checks are run as part of the PR checks, and so that the code climate report can be inspected for the PR.

**Review by markwhitfeld — CHANGES_REQUESTED** [review]
I have left a few comments here, but we can chat about the details on a call 👍

**Comment by markwhitfeld** [comment]
Here are some great articles that have some awesome recommendations, and security guidances for Github Actions:
- https://engineering.salesforce.com/github-actions-security-best-practices-b8f9df5c75f5/
- https://cycode.com/blog/github-actions-vulnerabilities/

File: .github/workflows/pr-validation.yml
Comment (rfprod) [reply .github/workflows/pr-validation.yml]: uncommented this step

File: .github/workflows/pr-validation.yml
Comment (rfprod) [reply .github/workflows/pr-validation.yml]: The comment was resolved during the call.

**Review by arturovt — COMMENTED** [review]
This looks great! I have a question: could you leave comments to commands like `find ~/.npmrc -exec shred -fuz {} +`? This would be obvious for reviewers on what this command is doing and will be easier to maintain it in the future, just because everyone will know what it's responsible for.
E.g. this would be great:
```sh
# This command is doing ..... `shred -fuz {}` is doing blah blah....
find ~/.npmrc -exec shred -fuz {} +
```

**Comment by rfprod** [comment]
> This looks great! I have a question: could you leave comments to commands like `find ~/.npmrc -exec shred -fuz {} +`? This would be obvious for reviewers on what this command is doing and will be easier to maintain it in the future, just because everyone will know what it's responsible for. E.g. this would be great:
> 
> ```shell
> # This command is doing ..... `shred -fuz {}` is doing blah blah....
> find ~/.npmrc -exec shred -fuz {} +
> ```

updated

File: .github/workflows/release.yml
Comment (markwhitfeld) [inline .github/workflows/release.yml:136]: We need to ensure that the artefacts from the `./@ngxs` folder created in the `build` job are present before this step so that they can be published.

File: .github/workflows/trunk.yml
Comment (markwhitfeld) [inline .github/workflows/trunk.yml:136]: Same here, we need to copy in the build artefacts before this step

**Review by markwhitfeld — CHANGES_REQUESTED** [review]

File: .github/workflows/release.yml
Comment (rfprod) [reply .github/workflows/release.yml:135]: what do you think of https://github.com/ngxs/store/pull/1852/commits/016b129c3afc993f28d3f337dd193a01a8980f63?

File: .github/workflows/trunk.yml
Comment (rfprod) [reply .github/workflows/trunk.yml:135]: same as above - https://github.com/ngxs/store/pull/1852/commits/016b129c3afc993f28d3f337dd193a01a8980f63

**Review by markwhitfeld — CHANGES_REQUESTED** [review]
Thank you so much for this!
Just a few more tweaks...
I'm still battling to get GitHub to allow this to run. Maybe you can experiment and test this in your own clean repo with some spoofed data and scripts. Then you are not blocked with not being able to run it in the meantime. What do you think?

File: .github/workflows/_setup.yml
Comment (rfprod) [reply .github/workflows/_setup.yml]: @markwhitfeld fixed

**Comment by markwhitfeld** [comment]
@rfprod after reading this article: https://github.blog/2021-11-29-github-actions-reusable-workflows-is-generally-available/ 
I am wondering if I am mistaken about reusable workflows, and that they are actually run in their own runner.
Potentially composite actions are what we need.
I think playing with this in a test repo is really critical to figuring out the best combination. We won't know until we try it.

File: .github/workflows/_setup.yml
Comment (rfprod) [reply .github/workflows/_setup.yml:64]: @markwhitfeld I think in this particular case the chance of something going wrong with the artifacts is extremely low.
Here's why:
1. when the workflow (trunk or release) is kicked off, the first job is `*-build` (`trunk-build` or `release-build`) <- `*-build` jobs refresh `workspace-cache` by rebuilding all packages and saving it to the cache.
2. Then several more jobs are executed: `*-test`, `*-integration-test`, `*-bundlesize`.
3. Then the `*-publish` job is executed. The `publish` job uses build artifacts that were generated by the `*-build` job (see point 1 of this list).

A possible attack surface is in between of `build` and `publish`, it's point 2 in the list above.
The attacker should be very persistent and determined to mess up the publish step.
If we assume that the workspace artifacts (including the build artifacts) are corrupted before `*-build`, they will be fixed after the `*-build` job completion.

I think, with the same probability someone could tap into the runner and corrupt the build artifacts. Of course there is a probability of this, but it tends to zero.

If you want an additional layer of protection to be sure that the build artifacts were not changed after caching and before publishing, we can hash the content of each directory with the build artifacts and save hashing results as part of the `workspace-cache`. The `_check-build-artifacts.yml` workflow can be configured to verify hashes in addition to ensuring that the directories exist.

**Comment by rfprod** [comment]
> @rfprod after reading this article: https://github.blog/2021-11-29-github-actions-reusable-workflows-is-generally-available/ I am wondering if I am mistaken about reusable workflows, and that they are actually run in their own runner. Potentially composite actions are what we need. I think playing with this in a test repo is really critical to figuring out the best combination. We won't know until we try it.

@markwhitfeld As I mentioned during our call, I haven't used reusable workflows like this before. :) I'll try if this approach actually works, maybe tomorrow.
As far as I know, each reusable workflow has its own isolated context in every sense, but maybe something has changed. The only way I am aware of to share artifacts between reusable workflows and not reusable workflows is [upload-artifact](https://github.com/actions/upload-artifact) + [download-artifact](https://github.com/actions/download-artifact). [actions/cache](https://github.com/actions/cache) is something that's generally used in workflows that are not reusable.

File: .github/workflows/pr-validation.yml
Comment (rfprod) [reply .github/workflows/pr-validation.yml:99]: @markwhitfeld `bundlesize` needs: `*-integration-test` and `*-integration-test` needs `*-build`. Artifacts should be there. Do you want verification that the direcoties exist (`_check-build-artifacts.yml` before `bundlesize`)?

Comment (rfprod) [reply .github/workflows/pr-validation.yml:99]: @markwhitfeld affed `app/integrations` dists to workspace cache as well

File: .github/workflows/_setup.yml
Comment (markwhitfeld) [reply .github/workflows/_setup.yml]: Thank you for your very well-thought-out response.
I think that it is good to bear this in mind, but I don't think it is essential right now.
I think that the simplest approach is to use the cache so that it is available to the next job, and then in the publish step, we just build it again. It is repeating the work, but everything preceding the publish leveraged the built files as part of the validation leading up to the publish. It is OK if the publish repeats a number of steps to be safer because it is on the tail end of the workflow and will not slow the PR workflow down.

File: .github/workflows/pr-validation.yml
Comment (markwhitfeld) [reply .github/workflows/pr-validation.yml:99]: Great!

File: .github/workflows/release.yml
Comment (markwhitfeld) [reply .github/workflows/release.yml:135]: I like it!

File: .github/workflows/pr-validation.yml
Comment (markwhitfeld) [reply .github/workflows/pr-validation.yml:99]: PS. Just FYI we would like to replace bundlesize with something else sometime soon.
If we have issues with bundlesize in this new workflow, we shouldn't spend too much effort to fix it.
See open issue: https://github.com/ngxs/store/issues/1764#issuecomment-927330712

**Review by markwhitfeld — APPROVED** [review]
Looking great.
Now just to see it running!!?!
Hopefully this approved review opens that option...
PS. Team, please wait for me to merge once we have potentially iterated a few times with seeing it in action.

**Comment by rfprod** [comment]
> Looking great. Now just to see it running!!?! Hopefully this approved review opens that option... PS. Team, please wait for me to merge once we have potentially iterated a few times with seeing it in action.

`pr-validation` can be tested in my fork ,see here https://github.com/rfprod/store/pull/1
Seems like it won't work in this repo before this PR is merged.
I am reviewing and addressing bugs in workflows to make it work.

**Comment by rfprod** [comment]
@markwhitfeld cache is immutable. Check my latest changes. `build` should go before **all** other jobs. Bundlesize wont work for integration tests unless `upload/download artifact` action is used in addition to workspace `cache` to pass additional artifacts to `bundlesize`. Jest cache won't persist between workflow runs as part of the workspace cache unless `test:ci` is executed in the same workflow as `build`.

https://github.com/rfprod/store/actions/runs/2485166750

**Comment by rfprod** [comment]
@markwhitfeld all jobs green.

Pipeline run https://github.com/rfprod/store/actions/runs/2491727586
Integration test artifacts https://github.com/ngxs/store/pull/1852/commits/b3ba642d1280dbd918babc9b266a7bd44ab0e0d0

File: .github/workflows/pr-validation.yml
Comment (rfprod) [reply .github/workflows/pr-validation.yml:26]: @markwhitfeld `fetch-depth: 0` stands for all history for all branches and tags. It should be a bit slower and will pull much more data than needed for other workflows than those that validate pull requests. From my experience with common repos and monorepos, it is generally not needed to pull all git history in the `trunk` or `release` workflows, because all tests that might require full commit history of all branches and tags (e.g. validate commit messages using commitlint) were done earlier during the corresponding PR validation.

If you think the default `fetch-depth: 0` makes sense, so be it.

Comment (markwhitfeld) [reply .github/workflows/pr-validation.yml:26]: Ok. Makes sense. Thanks!

Comment (rfprod) [reply .github/workflows/pr-validation.yml:26]: @markwhitfeld one more thing about the `actions/checkout` action.

Sources should be checked out prior to any steps that use composite actions like `./github/actions/...`, because relative references requires already checkout out sources.
If you want to wrap the step that check out sources as a separate composite action, the only way to reference this composite action that will work, if I understand correctly, is `ngxs/store/.github/actions/...`

Comment (markwhitfeld) [reply .github/workflows/pr-validation.yml:26]: Oh, hahaha. Yes, that makes total sense!

**Review by markwhitfeld — APPROVED** [review]

**Pull request merged**
